src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java

```diff
@@ -45,4 +48,21 @@ public Future<?> handle(String bucket, String filePath) {
 
     protected abstract Future<?> handle(ResourceDescription resource);
 
+    protected boolean hasReadAccess(String bucket, String filePath) {
+        String url = bucket + BlobStorageUtil.PATH_SEPARATOR + filePath;
+        return context.getApiKeyData().getAttachedFiles().contains(url);
+    }
+
+    protected boolean hasWriteAccess(String filePath, String decryptedBucket) {
+        String expectedUserBucket = BlobStorageUtil.buildUserBucket(context);
+        if (expectedUserBucket.equals(decryptedBucket)) {
+            return true;
+        }
+        String expectedAppDataBucket = BlobStorageUtil.buildAppDataBucket(context);
+        if (expectedAppDataBucket != null && expectedAppDataBucket.equals(decryptedBucket)) {
+            return filePath.startsWith(BlobStorageUtil.APPDATA_PATTERN.formatted(context.getSourceDeployment()));
```

Comment:
filePath is url encoded, you should also encode sourceDeployment for correct check

Reply:
fixed